### PR TITLE
Mark environments not freed before process exit persistent eagerly

### PR DESCRIPTION
**Pull request by Kha**
Another ~-10% on `import Mathlib`

### CODE REVIEW

**Comment by Kha** [comment]
!bench

File: src/Lean/Environment.lean [hunk 1]
```diff
@@ -765,7 +765,8 @@ partial def importModulesCore (imports : Array Import) : ImportStateM Unit := do
       moduleNames := s.moduleNames.push i.module
     }
 
-def finalizeImport (s : ImportState) (imports : Array Import) (opts : Options) (trustLevel : UInt32 := 0) : IO Environment := do
+def finalizeImport (s : ImportState) (imports : Array Import) (opts : Options) (trustLevel : UInt32 := 0)
```
Comment (leodemoura) [inline src/Lean/Environment.lean:768]: Please add documentation for `leakEnv`

Comment (leodemoura) [reply src/Lean/Environment.lean:768]: And motivate why this was added. We will forget the discussion we had this week.

File: src/Lean/Environment.lean [hunk 2]
```diff
@@ -797,18 +798,28 @@ def finalizeImport (s : ImportState) (imports : Array Import) (opts : Options) (
       moduleData   := s.moduleData
     }
   }
-  let env ← setImportedEntries env s.moduleData
-  let env ← finalizePersistentExtensions env s.moduleData opts
+  env ← setImportedEntries env s.moduleData
+  if leakEnv then
+    /- When allowed to leak `env`, we eagerly mark it persistent here, which
+       avoids RC updates as well as costly MT markings when e.g. an interpreter
+       closure (wich contains the environment) is put in an `IO.Ref` like would
```
Comment (leodemoura) [inline src/Lean/Environment.lean:805]: wich => which

File: src/Lean/Environment.lean [hunk 3]
```diff
@@ -797,18 +798,28 @@ def finalizeImport (s : ImportState) (imports : Array Import) (opts : Options) (
...
+       happen with user environment extensions loaded in the next step. -/
+    env := Runtime.markPersistent env
+  env ← finalizePersistentExtensions env s.moduleData opts
+  if leakEnv then
+    /- Also mark environment extension states. -/
+    env := Runtime.markPersistent env
```
Comment (leodemoura) [inline src/Lean/Environment.lean:811]: You are calling `markPersistent` twice. It is worth explaining. I am guessing that you are concerned some extension may mark the environment as `MT` at `finalizePersistentExtensions`.

File: src/Lean/Environment.lean (hunk 1 above)
Comment (Kha) [reply src/Lean/Environment.lean:777]: Yes, I wasn't sure where to document it. I've split the comment below, moved the first part into the docstring, and extended comments in general.

File: src/Lean/Environment.lean (hunk 3 above)
Comment (Kha) [reply src/Lean/Environment.lean:825]: That could happen; mostly I just want to give a consistent guarantee about what is returned with `leakEnv`.

**Comment by Kha** [comment]
The `import Lean` regression is interesting, it shows it is not quite like `import Mathlib` after all: it doesn't have any interpreted code of course so before this PR, the environment object stays ST and marking it persistent is additional overhead. But as we want to introduce parallelism in basically any file, and having the environment be MT is definitely worse than persistent, I think this regression is acceptable. @leodemoura Do you agree?

I double-checked the `import Mathlib` speedup on top of https://github.com/leanprover-community/mathlib4/pull/8199:
```
Benchmark 1: lean Mathlib.lean
  Time (mean ± σ):      1.464 s ±  0.032 s    [User: 1.145 s, System: 0.311 s]
  Range (min … max):    1.437 s …  1.544 s    10 runs

Benchmark 2: lean +lean4 Mathlib.lean
  Time (mean ± σ):      1.336 s ±  0.011 s    [User: 1.041 s, System: 0.294 s]
  Range (min … max):    1.319 s …  1.355 s    10 runs

Summary
  'lean +lean4 Mathlib.lean' ran
    1.10 ± 0.03 times faster than 'lean Mathlib.lean'
    Benchmark 1: lean Mathlib.lean
  Time (mean ± σ):      1.464 s ±  0.032 s    [User: 1.145 s, System: 0.311 s]
  Range (min … max):    1.437 s …  1.544 s    10 runs

Benchmark 2: lean +lean4 Mathlib.lean
  Time (mean ± σ):      1.336 s ±  0.011 s    [User: 1.041 s, System: 0.294 s]
  Range (min … max):    1.319 s …  1.355 s    10 runs

Summary
  'lean +lean4 Mathlib.lean' ran
    1.10 ± 0.03 times faster than 'lean Mathlib.lean'
```

**Comment by Kha** [comment]
> having the environment be MT is definitely worse than persistent

If we want to 100% make sure about this, we can also postpone the PR until such time

File: src/Init/Util.lean [hunk 4]
```diff
@@ -85,3 +85,22 @@ private def outOfBounds [Inhabited α] : α :=
...
+  marked single-threaded. This will make reference counter updates atomic and
+  thus more costly. It can still be useful to do eagerly when the value will be
+  shared between threads later anyway and there is available time budget to mark
+  it now. -/
+@[extern "lean_runtime_mark_multi_threaded"]
+def Runtime.markMultiThreaded (a : α) : α := a
+
+/--
+  Marks given value and its object graph closure as multi-threaded if currently
+  marked single-threaded. This will remove reference counter updates but prevent
```
Comment (eric-wieser) [inline src/Init/Util.lean:100]: Is this first sentence deliberately identical to the one in the other docstring?

Comment (Kha) [reply src/Init/Util.lean:100]: Wow, thanks, I must have overwritten it in a later revision

**Comment by leodemoura** [comment]
> But as we want to introduce parallelism in basically any file, and having the environment be MT is definitely worse than persistent, I think this regression is acceptable as it should not affect real-life files. @leodemoura Do you agree?

Agreed. Let's merge the PR.
Before merging, please include your analysis as a comment in the file. We don't need the actual numbers.

**Review by leodemoura — APPROVED** [review]

**Pull request merged**